File: TypeScript/Algorithm/Algorithm I/189. Rotate Array.ts
/**
 * Rotates a portion of an array between the start and end indices.
 * @param nums - The array of numbers to be rotated.
 * @param startIndex - The starting index of the rotation.
 * @param endIndex - The ending index of the rotation.
 */
function rotateArray(nums: number[], startIndex: number, endIndex: number): void {
  while (startIndex < endIndex) {
    const tempValue: number = nums[startIndex];

    nums[startIndex] = nums[endIndex];
    nums[endIndex] = tempValue;

    startIndex++;
    endIndex--;
  }
}

/**
 * Rotates the elements of an array to the right by a specified number of steps.
 * @param nums - The array of numbers to be rotated.
 * @param k - The number of steps to rotate the array.
 */
function rotate(nums: number[], k: number): void {
  const LENGTH: number = nums.length;

  k = k % LENGTH;

  rotateArray(nums, 0, LENGTH - 1);
  rotateArray(nums, 0, k - 1);
  rotateArray(nums, k, LENGTH - 1);
}

File: Java/Algorithm/Algorithm I/189. Rotate Array.java
class Solution {

  /**
   * Rotates a portion of an array in-place.
   *
   * @param nums       The array of integers to be rotated.
   * @param startIndex The starting index of the portion to rotate.
   * @param endIndex   The ending index of the portion to rotate.
   */
  private void arrayRotate(int[] nums, int startIndex, int endIndex) {
    while (startIndex < endIndex) {
      int tempValue = nums[startIndex];

      nums[startIndex] = nums[endIndex];
      nums[endIndex] = tempValue;

      startIndex++;
      endIndex--;
    }
  }

  /**
   * Rotates an array to the right by a specified number of steps.
   *
   * @param nums The array of integers to be rotated.
   * @param k    The number of steps to rotate the array to the right.
   */
  public void rotate(int[] nums, int k) {
    int NUMS_LENGTH = nums.length;

    k = k % NUMS_LENGTH;

    this.arrayRotate(nums, 0, NUMS_LENGTH - 1);
    this.arrayRotate(nums, k, NUMS_LENGTH - 1);
    this.arrayRotate(nums, 0, k - 1);
  }
}
